noether_simulator/src/thickness_simulator_node.cpp

```diff
@@ -0,0 +1,247 @@
+  noether_msgs::ThickSimulatorResult result_;
+
+  tool_path_planner::ProcessPath convertPoseArraytoVTK(geometry_msgs::PoseArray array)
+  {
+    tool_path_planner::ProcessPath path;
+
+    vtkSmartPointer<vtkPolyData> line = vtkSmartPointer<vtkPolyData>::New();
+    vtkSmartPointer<vtkPolyData> derivatives_ = vtkSmartPointer<vtkPolyData>::New();
+    vtkSmartPointer<vtkPoints> pts = vtkSmartPointer<vtkPoints>::New();
+
+    // set the point positions
+    for(int ii = 0; ii < array.poses.size(); ii++)
+    {
+      geometry_msgs::Pose pose = array.poses[ii];
+
+      double pt[3];
+      pt[0] = (double)pose.position.x;
+      pt[1] = (double)pose.position.y;
+      pt[2] = (double)pose.position.z;
+
+      pts->InsertNextPoint(pt);
+    }
+
+    line->SetPoints(pts);
+    derivatives_->SetPoints(pts);
+
+    // calculate point normals
+    vtkSmartPointer<vtkDoubleArray> normals = vtkSmartPointer<vtkDoubleArray>::New();
+    vtkSmartPointer<vtkDoubleArray> deriv_normals = vtkSmartPointer<vtkDoubleArray>::New();
+    normals->SetNumberOfComponents(3);
+    deriv_normals->SetNumberOfComponents(3);
+
+    for(int ii = 0; ii < line->GetPoints()->GetNumberOfPoints(); ii++)
+    {
+      double pt[3];
+      geometry_msgs::Quaternion q = array.poses[ii].orientation;
+
+      //calculate z-axis from quaternion
+      double x = (2*q.x*q.z + 2*q.y*q.w);
+      double y = (2*q.y*q.z - 2*q.x*q.w);
+      double z = (1 - 2*q.x*q.x - 2*q.y*q.y);
+      double n[3] = {x, y, z};
+      normals->InsertNextTypedTuple(n);
+
+      // calculate y-axis from quaternion
+      x = (2*q.x*q.y - 2*q.z*q.w);
+      y = (1 - 2*q.x*q.x - 2*q.z*q.z);
+      z = (1 - 2*q.y*q.z + 2*q.x*q.w);
+      double m[3] = {x, y, z};
+      deriv_normals->InsertNextTypedTuple(m);
+    }
+    line->GetPointData()->SetNormals(normals);
+    path.line = line;
+
+    derivatives_->GetPointData()->SetNormals(deriv_normals);
+    path.derivatives = derivatives_;
+
+    vtkSmartPointer<vtkPolyData> inter_ = vtkSmartPointer<vtkPolyData>::New();
+    vtkSmartPointer<vtkParametricSpline> spline_ = vtkSmartPointer<vtkParametricSpline>::New();
+    path.intersection_plane = inter_;
+    path.spline = spline_;
+
+    return path;
+  }
+
+public:
+
+  ProcessSimulatorNode(std::string name) :
+    simulation_service_(nh_, name, boost::bind(&ProcessSimulatorNode::executeCB,this, _1), false),
+    tool_height_(2),
+    process_rate_(0),//not used in utest
+    display_sigma_(0),//not used in utest
+    tool_radius_(1)
+  {
+    simulation_service_.start();
+  }
+  void executeCB(const noether_msgs::ThickSimulatorGoalConstPtr &goal)
+  {
+    ros::Rate r(0.05);
+    bool success = true;
+    feedback_.percent = 0.0;
+    ROS_INFO("simulator running");
+    int num_meshes= goal->input_mesh.size();
+    int num_paths = goal->path.size();
+    noether_simulator::NoetherSimulator sim;
+
+    //add tool
+    tool_path_planner::ProcessTool tool;
+    tool.pt_spacing = 0.5;
```

Comment:
get all these parameters from the ros parameter server, but use these values as default if not set on ros-p-server